### PR TITLE
[Google Blockly] Support JSON with loading blocks and Flappy/Bounce block arrangements (Save sources in JSON, pt 2)

### PR DESCRIPTION
**Pull request by mikeharv**
## Context
We have always saved student Blockly code as an XML string. While XML is still supported in Google Blockly, JSON is the preferred format and will allow us to eventually store function definitions off of the main workspace.
To accomplish this, we can create helper functions in each Blockly wrapper that will handle saving and loading student code in whichever manner is appropriate. We can change the format of this source without breaking existing projects.

## Strategy
We will not perform a bulk update on existing projects as the XML sources should continue to work. Instead, we will changes things so that the next time a Google Blockly project is opened we convert it to JSON. This overall strategy was established by a previous PR:
1. Replace all* instances where student code is saved with a new helper that matches in both wrappers:
 --  https://github.com/code-dot-org/code-dot-org/pull/50786
3. Replace where student code is loaded (in StudioApp.js) with a new util helper that matches in both wrappers. (This PR)
4. Update both util helpers in the Google Blockly space to start working with JSON, behind an experiment flag. (This PR)

In StudioApp.js, we find a general flow for adding the blocks to the workspace:
[`setStartBlocks`](https://github.com/code-dot-org/code-dot-org/blob/mike/load-blocks-from-json-experiment/apps/src/StudioApp.js#L2699-L2736) - This pulls down the source for the blocks, usually in XML. It then calls:
  -- [`arrangeBlockPosition`](https://github.com/code-dot-org/code-dot-org/pull/50813/files#diff-dd5a7531bc734452cd21d4bd0ab327de7fb4a8a763a29864898ae369fb081e90L1265-L1291) - Rewrites the x/y position of some blocks (only used by Flappy/Bounce - Google Blockly labs)
  -- [`loadBlocks`](https://github.com/code-dot-org/code-dot-org/pull/50813/files#diff-dd5a7531bc734452cd21d4bd0ab327de7fb4a8a763a29864898ae369fb081e90L1247-L1250) - Actually loads the blocks onto the workspace.

### Update how we load blocks

We are moving to a place where the `startBlocks` parameter of `setStartBlocks_` could be either stringified JSON or XML for a Blockly lab. We don't need to actually modify anything here, but it may be worth looking through `setStartBlocks_`, linked above, to understand how it works.

As we will be supporting JSON or XML, `arrangeBlockPosition` must work in either case. A `try...catch` is employed to first attempt modification assuming `startBlocks` is JSON. The catch handles any error by assuming that `startBlocks` must then be XML. Either way, we return an updated source of truth of the blocks. Two new util functions are created for this purpose. `arrangeBlockPosition` isn't really needed for any CDO Blockly labs, so the CDO version that was added to that corresponding wrapping can just return its first argument immediately.

`loadBlocks` must also be updated to support JSON. We can employ a similar `try...catch` structure to first attempt loading the blocks from JSON and fall back to XML if that doesn't work. The existing method for loading blocks from XML works with both Blockly versions. This is also the only place we seem to load blocks, so nothing here is moved out to a util. 

### Allow saving as JSON via experiment
It was relatively light touch to also add an experiment flag that tells our Google Blockly labs to start saving code changes as JSON. This would allow testing the new flow on production on an opt-in basis. Production is the best environment for testing this as adhocs and localhost don't have the same volume of projects on different accounts.

This video demonstrates a potential flow for testing that we are now supporting both JSON and XML, using Bounce:

https://user-images.githubusercontent.com/43474485/225768429-2e81408a-acc3-4752-a195-8091ff314ec4.mp4

Video beats:
- 0:00 - A block is added to a new bounce project.
- 0:02 - Project saved as XML when Run button pressed.
- 0:08 - Project refreshed to show changes persist
- 0:14 - Experiment flag added to url
- 0:17 - Project reloads, still loading blocks from previous XML. Console immediately shows the same blocks as JSON
- 0:24 - Project saved as JSON and output to the console
- 0:47 - Changes persist on further reload. Console indicates "JSON used to load blocks"
- 1:03 - Experiment flag removed
- 1:09 - Project reloads. Source is still JSON, which is used to load blocks. (This is expected. We don't want projects to break just because the experiment is disabled.)
- 1:16 - A final change is made to the project, which is saved in XML.
- 1:24 - The project reloads a final time. This time the source is back in XML.

In summary, this video demonstrates that the new utils enable us to load an XML project, convert it to JSON, and, if need be, convert it back. At all steps along the way, the blocks are loaded as expected.

I also tested this flow in a CDO Blockly lab (Artist). No matter the state of the experiment flag, XML is still always used for loading and saving blocks in CDO Blockly labs.

## Links

Jira - https://codedotorg.atlassian.net/browse/SL-640

## Follow-up work

Add tests. For example, we should test that loading a set of XML blocks with Google Blockly and then saving it yields an expected JSON output. We could also test that loading blocks from either type of source leads to the correct set of blocks being created. 

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: apps/src/StudioApp.js [hunk 1]
```diff
@@ -1228,7 +1228,7 @@ StudioApp.prototype.showNextHint = function() {
 /**
  * Initialize Blockly for a readonly iframe.  Called on page load. No sounds.
  * XML argument may be generated from the console with:
```
Comment (bencodeorg) [inline apps/src/StudioApp.js]: Nit: with experiment enabled this would generate JSON yeah?

File: apps/src/blockly/addons/cdoUtils.js [hunk 2]
```diff
@@ -104,7 +106,56 @@ export function getUserTheme(themeOption) {
...
+  if (experiments.isEnabled(experiments.BLOCKLY_JSON)) {
+    const blocksJson = Blockly.serialization.workspaces.save(workspace);
+    console.log('blocks saved as JSON:', blocksJson);
+    return JSON.stringify(blocksJson);
+  } else {
+    return Blockly.Xml.domToText(Blockly.Xml.blockSpaceToDom(workspace));
+  }
+}
+
+export function arrangeBlocksJson(startBlocksJson, arrangement) {
```
Comment (bencodeorg) [inline apps/src/blockly/addons/cdoUtils.js]: Finding this a little hard to follow...a couple questions:
- why is block defined outside the loop?
- could you assign block.type (type) to a variable for readability
- i think? the point of this is to just modify each block in `code.blocks.blocks` -- could your loop just be `code.blocks.blocks.forEach(block => do stuff)`? Looks like this is modeled off the other function (arrangeBlocksXml), which maybe was written before forEach existed.

I see you have testing in "next steps", but I wonder if it would be helpful in review to have unit tests here to help understand the expected inputs/outputs of this function, especially since this is probably a pretty important function :).

File: apps/src/StudioApp.js [hunk 3]
```diff
@@ -1242,11 +1242,20 @@ StudioApp.prototype.initReadonly = function(options) {
...
 /**
  * Load the editor with blocks.
- * @param {string} blocksXml Text representation of blocks.
+ * @param {string} blocks Text representation of blocks (XML or JSON).
  */
-StudioApp.prototype.loadBlocks = function(blocksXml) {
-  var xml = parseXmlElement(blocksXml);
-  Blockly.Xml.domToBlockSpace(Blockly.mainBlockSpace, xml);
+StudioApp.prototype.loadBlocks = function(blocks) {
+  try {
```
Comment (bencodeorg) [inline apps/src/StudioApp.js:1248]: try/catch here and below feels odd...can we be explicit about whether we expect blocks to be xml or json?

**Review by bencodeorg — COMMENTED** [review]
Very cool to see swapping back and forth between XML and JSON working! A couple style questions but looking good

File: apps/src/StudioApp.js (hunk 1 above)
Comment (mikeharv) [reply apps/src/StudioApp.js]: True. Updating this comment was more related to the previous PR where we implemented this Util: https://github.com/code-dot-org/code-dot-org/pull/50786

I'm actually not sure what the value of this comment is. For now, read only workspaces are always going to be sourced with XML, so maybe it makes sense to leave the comment instructions as they were, using the XML methods.

File: apps/src/StudioApp.js (hunk 3 above)
Comment (fisher-alice) [reply apps/src/StudioApp.js:1248]: Can we add a condition to check if json/xml? maybe add an isJson helper function?

File: apps/src/blockly/addons/cdoUtils.js [hunk 4]
```diff
@@ -104,7 +106,83 @@ export function getUserTheme(themeOption) {
...
+      if (arrangement[type].y && !block.y) {
+        block.y = arrangement[type].y;
+      }
+      console.log(`Applied arrangement for block of type ${block.type}`);
+    }
+  });
+  return JSON.stringify(code);
+}
+
+function arrangeBlocksXml(startBlocksXml, arrangement) {
```
Comment (fisher-alice) [inline apps/src/blockly/addons/cdoUtils.js:164]: More readable than previous version 😄
